[WIP] dnf gracefully handle installed excluded pkg install operation

maxamillion opened the pull request:
Signed-off-by: Adam Miller <admiller@redhat.com>

##### SUMMARY
dnf install attempt on an already installed package that is excluded no longer causes failure

Fixes #33187

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
dnf

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.8.0.dev0 (bugfix/33187-dnf-excludes-failure 25990cf1b4) last updated 2018/09/25 15:46:19 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/admiller/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/admiller/src/dev/ansible/lib/ansible
  executable location = /home/admiller/src/dev/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

maxamillion committed:
dnf gracefully handle installed excluded pkg install operation

Fixes: 33187

Signed-off-by: Adam Miller <admiller@redhat.com>

agaffney commented on lib/ansible/modules/packaging/os/dnf.py:
```diff
@@ -636,6 +636,14 @@ def _mark_package_install(self, pkg_spec, upgrade=False):
         is_newer_version_installed = self._is_newer_version_installed(pkg_spec)
         is_installed = self._is_installed(pkg_spec)
         try:
+            if (pkg_spec in self.base.conf.exclude) or ('*' in self.base.conf.exclude):
```
I'm not sure how it works with `dnf`, but with `yum`, the name of the package you're trying to install doesn't necessarily need to be an exact match for what's in `exclude` for the package to be excluded. It's better than nothing, though.

maxamillion replied:
@agaffney that's fair, I could try doing some heuristics with `fnmatch`. Thoughts?

agaffney replied:
Is there some built-in method for comparing the `pkg_spec` against an `exclude` entry with a wildcard? There's obviously support for it somewhere in `dnf`, but it may or may not be exposed.

maxamillion replied:
I was looking though the [DNF API](https://dnf.readthedocs.io/en/latest/api_queries.html#dnf.query.Query) and I didn't find anything, I also looked through the DNF code itself on how it implements this functionality and it's making API calls that are exposed but are not compatible with older versions of DNF so I'm not sure how I would handle this reliably in a way that would be compatible across DNF v1/v2/v3. I'm still looking though, will update as I find more.

maxamillion committed:
handle dnf excludes gracefully with dnf api mechanisms

This works on all non-EOL Fedora releases, but will fail the tests
because at the time of this commit, the versions Ansible CI runs
against are 3 and 4 releases EOL, respectively.

The issue with the older versions of Fedora is that hawkey segfaults
attempting to refill the dnf sack after we've extracted the excludes
information and reset the config to the original state.

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion commented:
@agaffney so I put together a new commit and I'm marking this as `[WIP]` until I can sort out how to work around hawkey segfaulting on older Fedora releases, but it works fine on Fedora 27 and 28. However, I'm not sure I like the fix because it essentially doubles the runtime of the module because the bulk of the time spent is doing the `base.fill_sack()` operation and that has to be done twice to extract the expanded excludes packages.

maxamillion commented:
For reference.

Patched version with this PR's code included.
```
$ time ansible localhost -m dnf -a 'name=erlang-tools state=installed' 
localhost | SUCCESS => {
    "changed": false,
    "msg": "Nothing to do",
    "rc": 0,
    "results": [
        "Not installed due to excludes: erlang-tools"
    ]
}

real    0m5.890s
user    0m4.318s
sys     0m0.428s
```

Current devel branch
```
$ time ansible localhost -m dnf -a 'name=erlang-tools state=installed'
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

localhost | FAILED! => {
    "changed": false,
    "failures": [
        "No package erlang-tools available."
    ],
    "msg": "Failed to install some of the specified packages",
    "rc": 1,
    "results": []
}

real    0m2.652s
user    0m2.399s
sys     0m0.274s
```

maxamillion commented:
Closing, the dnf command line doesn't gracefully handle this either.

```
$ dnf install -x erlang-tools erlang-tools
Last metadata expiration check: 0:00:08 ago on Fri 28 Sep 2018 09:11:11 AM CDT.
No match for argument: erlang-tools
Error: Unable to find a match
```

abadger commented on lib/ansible/modules/packaging/os/dnf.py:
```diff
@@ -303,6 +304,30 @@ def __init__(self, module):
+    def _pattern_to_pkgname(self, base, pattern):
+        if dnf.util.is_glob_pattern(pattern):
+            q = base.sack.query().filter(name__glob=pattern)
+            return map(lambda p: p.name, q)
```
Anytime I reach for map or filter, I like to use a list comprehension of generator exception instead as it is faster and considered more pythonic.
``` python
return [p.name for p in q]
```

abadger commented on lib/ansible/modules/packaging/os/dnf.py:
```diff
@@ -303,6 +304,30 @@ def __init__(self, module):
+            self.base.conf.disable_excludes = []
+        self.base.fill_sack()
+
+        return itertools.chain.from_iterable(expanded_excludes)
```
Going through itertools here ends up creating both a list and a generator so it's likely slower than just returning a list.  If the list is very large and your goal is to reduce memory consumption, you can use a generator expresssion instead of a list comprehension on line 322:
``` python
expanded_excludes = (self._pattern_to_pkgname(self.base, p) for p in self.base.conf.exclude)
```

However, for iteration over most data, lists are faster than other iterables so it is a time-space tradeoff.

maxamillion replied:
Good to know, also this ends up needing to be a list anyways because once the `dnf.query.Query` object goes out of scope, the generator returns nothing. Thanks!
